### PR TITLE
Extract pod name generation functions

### PR DESCRIPTION
**Pull request by cronik**
Extract pod name generation functions to PodUtils for better testability and
reusability.

> This pull request is part of a series of small changes and enhancements to support a new plugin to add ephemeral container support to Kubernetes agents. This feature was originally built as an enhancement to this plugin and has been running in live production environment for a few years. CloudBees engineers suggested we create a new plugin rather than contribute here. https://github.com/jenkinsci/kubernetes-plugin/compare/master...cronik:kubernetes-plugin:feature/ephemeral-containers

### Testing done

Added unit tests and have been running in production environment for a few years.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests that demonstrate the feature works or the issue is fixed

### CODE REVIEW

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java [hunk 1]
```diff
@@ -57,18 +65,85 @@ public static List<ContainerStatus> getWaitingContainers(Pod pod) {
...
+            return Optional.empty();
+        }
+
+        Optional<String> containerWorkingDir = pod.getSpec().getContainers().stream()
+                .filter(c -> StringUtils.equals(c.getName(), containerName))
+                .findAny()
+                .map(Container::getWorkingDir);
+
+        if (containerWorkingDir.isEmpty()) {
+            containerWorkingDir = pod.getSpec().getEphemeralContainers().stream()
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java:138]: Even if this will probably work (as it's looking into ephemeral containers only when there is no other containers), I think this is not right in terms of design and isolation. It should use the Jenkins extension system to hook functionality in here. Maybe something like `PodContainerSource`:

```java
public abstract class PodContainerSource implements ExtensionPoint {
    public abstract Optional<String> getContainerWorkingDir(Pod pod, String containerName);
    
    public static List<PodContainerSource> all() {
        return ExtensionList.lookup(PodContainerSource.class);
    }
}
```

Then create an implementation in this plugin:

```java
@Extension(ordinal = Double.MAX_VALUE) // this extension goes always first
public static final class DefaultPodContainerSource extends PodContainerSource {

    @Override
    public Optional<String> getContainerWorkingDir(Pod pod, String containerName) {
        return pod.getSpec().getContainers().stream()
                    .filter(c -> Objects.equals(c.getName(), containerName))
                    .findAny()
                    .map(Container::getWorkingDir);
    }
}
```

And finally, `PodUtils#getContainerWorkingDir` would be:

```java
public static Optional<String> getContainerWorkingDir(Pod pod, String containerName) {
    return PodContainerSource.all().stream().filter(cs -> cs.getContainerWorkingDir(pod, containerName)).findFirst();
}
```

Then you can create a second implementation in the other plugin to look for ephemeral containers.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java [hunk 2]
```diff
@@ -57,18 +65,85 @@ public static List<ContainerStatus> getWaitingContainers(Pod pod) {
...
         return getContainerStatus(pod).stream().filter(predicate).collect(Collectors.toList());
     }
 
+    /**
+     * Lookup container status (either main container or ephemeral container).
+     * @param pod pod resource to inspect
+     * @param containerName container to locate
+     * @return container status if found
+     */
+    public static Optional<ContainerStatus> getContainerStatus(Pod pod, String containerName) {
```
Comment (amuniz) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java:98]: This is not called from anywhere in this plugin. If it's required for the extension plugin, move it there.

Comment (cronik) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java:98]: It will be called in a future PR to that improves the KuberenetesCloud log handling. I can move it to that PR but it was easier to cherry pick it here since I add unit tests for PodUtils which has none before.

https://github.com/cronik/kubernetes-plugin/blob/f4c309255237b3db9fde9084ec126e293cbea984/src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesComputer.java#L167

Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java:98]: I agree with Antonio, this will be easier for later forensics (if any)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java (hunk 1 above)
Comment (cronik) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodUtils.java]: I've reduced the scope of this PR and will work on this in a new one.

**Review by amuniz — APPROVED** [review]
